### PR TITLE
*: split index region with lower upper syntax

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 

Add new syntax for `split index region`.

```sql
SPLIT TABLE [table_name] INDEX [index_name] BETWEEN (lower_value...) AND (upper_value...) REGIONS [region_num]
```

#### Attention

- The `lower_value` and `upper_value` value count should less than the index column num.
- The `lower_value` and `upper_value` value count should more than 0.
- The max split index region num currently is `1000`. This limit in TiDB code. 

**Eg1: int type.**

```sql
split table `t` index `idx1` between (0) and (1000) regions 5;
```

Upper sql will split `5` regions of the table `t` index `idx1` in range `0~1000`. Every region range like below:

Region1: -inf ~ 200

Region2: 200 ~ 400

Region3: 400 ~ 600

Region4: 600 ~ 800

Region5: 800 ~ +inf

**Eg2: varchar type**

```sql
split table `t` index `idx1` between ('a') and ('z') regions 26;
```

Upper sql will split `26` regions. 

Region1:  -inf ~ b

Region2:  b ~ c

...

Region26: y ~ +inf

**Eg3: time type.**

```sql
split table `t` index `idx1` between ('2010-01-01 00:00:00') and ('2020-01-01 00:00:00') regions 10;
```

Upper sql will split `10` regions. 

Region1:  -inf                               ~  2011-01-01 00:00:00
Region2:  2011-01-01 00:00:00 ~ 2012-01-01 00:00:00

...

Region10:  2011-01-01 00:00:00 ~ +inf

**Eg4: multi-column index. `idx1 (col1,col2,col3)`. **

- if you only want to split index region by first column `col1`. You can use SQL like Eg1,Eg2,Eg3. Suppose col1 is int type.

  ```sql
  split table `t` index `idx1` between (-1000) and (1000) regions 5;
  ```

- If you wanto split index regions by first two column. You can use SQL like below.

  ```sql
  split table `t` index `idx1` between (1000, "a") and (1000,"z") regions 26;
  ```

  

### What is changed and how it works?

Related Parser PR: https://github.com/pingcap/parser/pull/321

Split `num` regions between `lower` and `upper` value. 

Because the key value is byte slice, I cann't do this simple like ` (max-min) / num` .

To solve this problem：

1. find the longest common prefix of min and max value. 
2. get 8 diff-bytes of min and max from begin, if less than 8, fill with `0xff` 
3. convert 8 diff-bytes to uint64. Then use the `diff_uint64/num` to get the `step` value. 

### Check List 

Tests 

- Unit test
- Integration test

Code changes

- Has exported function/method change

Side effects

Related changes

### CODE REVIEW

**Comment by crazycs520** [comment]
@winkyao @zimulala @nolouch PTAL

File: executor/split.go [hunk 1]
```diff
@@ -75,10 +79,125 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+	i := 0
+	for ; i < l; i++ {
+		if s1[i] != s2[i] {
+			break
+		}
+	}
+	return i
+}
+
+func getDiffBytesValue(startIdx int, min, max []byte) uint64 {
```
Comment (winkyao) [inline executor/split.go]: Please add some comments

File: executor/split.go [hunk 2]
```diff
@@ -75,10 +79,125 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+		startValueTemp = startValueTemp[:8]
+	}
+	startValue := make([]byte, 0, 8)
+	startValue = append(startValue, startValueTemp...)
+	for i := len(startValue); i < 8; i++ {
+		startValue = append(startValue, 0)
+	}
+	startV := binary.BigEndian.Uint64(startValue)
+	// To get `num` regions, only need to split `num-1` idx keys.
+	num--
```
Comment (winkyao) [inline executor/split.go]: remove `num--` and use `for i := 0; i < num - 1; i++`

File: executor/split.go [hunk 3]
```diff
@@ -75,10 +79,125 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+
+	for i := len(diff); i < 8; i++ {
+		diff = append(diff, 0xff)
+	}
+	diffValue := binary.BigEndian.Uint64(diff)
+	return diffValue
+}
+
+func getValuesList(min, max []byte, num int, valuesList [][]byte) [][]byte {
+	startIdx := longestCommonPrefixLen(min, max)
```
Comment (winkyao) [inline executor/split.go]: startIdx --> commonPrefixIdx

File: executor/split.go [hunk 4]
```diff
@@ -75,10 +79,125 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+	diffValue := getDiffBytesValue(startIdx, min, max)
```
Comment (winkyao) [inline executor/split.go]: `diffValue` --> `diffRangeOfMinMax`

File: executor/split.go [hunk 5]
```diff
@@ -75,10 +79,117 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+		buf = make([]byte, 0, 8)
+		buf = append(buf, bs...)
+		for i := len(buf); i < 8; i++ {
+			buf = append(buf, pad)
+		}
+	}
+	return binary.BigEndian.Uint64(buf)
+}
+
+// getValuesList use to get `num` values between min and max value.
```
Comment (winkyao) [inline executor/split.go]: is used to

File: executor/split_test.go [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+		{[]byte{0}, []byte{128}, 0, binary.BigEndian.Uint64([]byte{128, 255, 255, 255, 255, 255, 255, 255})},
+		{[]byte{'a'}, []byte{'z'}, 0, binary.BigEndian.Uint64([]byte{'z' - 'a', 255, 255, 255, 255, 255, 255, 255})},
+		{[]byte("abc"), []byte{'z'}, 0, binary.BigEndian.Uint64([]byte{'z' - 'a', 255 - 'b', 255 - 'c', 255, 255, 255, 255, 255})},
+		{[]byte("abc"), []byte("xyz"), 0, binary.BigEndian.Uint64([]byte{'x' - 'a', 'y' - 'b', 'z' - 'c', 255, 255, 255, 255, 255})},
+		{[]byte("abc"), []byte("axyz"), 1, binary.BigEndian.Uint64([]byte{'x' - 'b', 'y' - 'c', 'z', 255, 255, 255, 255, 255})},
+		{[]byte("abc0123456"), []byte("xyz01234"), 0, binary.BigEndian.Uint64([]byte{'x' - 'a', 'y' - 'b', 'z' - 'c', 0, 0, 0, 0, 0})},
+	}
+
+	for _, ca := range cases {
+		l0 := longestCommonPrefixLen(ca.min, ca.max)
```
Comment (winkyao) [inline executor/split_test.go]: `10`?

File: planner/core/planbuilder.go [hunk 7]
```diff
@@ -1607,44 +1608,97 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+	schema := expression.TableInfo2SchemaWithDBName(b.ctx, node.Table.Schema, tblInfo)
+	mockTablePlan.SetSchema(schema)
 
-	indexValues := make([][]types.Datum, 0, len(node.ValueLists))
-	for i, valuesItem := range node.ValueLists {
-		if len(valuesItem) > len(indexInfo.Columns) {
-			return nil, ErrWrongValueCountOnRow.GenWithStackByArgs(i + 1)
-		}
-		valueList := make([]types.Datum, 0, len(valuesItem))
+	checkValue := func(valuesItem []ast.ExprNode) ([]types.Datum, error) {
```
Comment (winkyao) [inline planner/core/planbuilder.go]: This anonymous function name is `checkValue`, but it just converts the value from exprNode to Datum. Maybe better to name it `convertValue2ColumnType`

File: planner/core/planbuilder.go [hunk 8]
```diff
@@ -1607,44 +1608,97 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+				return nil, err
+			}
+			indexValues = append(indexValues, values)
+		}
+		p.ValueLists = indexValues
+		return p, nil
+	}
 
-			valueList = append(valueList, value)
+	checkMinMaxValue := func(valuesItem []ast.ExprNode, name string) ([]types.Datum, error) {
```
Comment (winkyao) [inline planner/core/planbuilder.go]: Please use two functions to handle (min,max) syntax and valuesList syntax severally. To make your code cleaner.

File: planner/core/planbuilder.go [hunk 9]
```diff
@@ -1607,44 +1608,97 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
-		Table:      tableInPlan,
-		IndexInfo:  indexInfo,
-		ValueLists: indexValues,
-	}, nil
+	maxValue, err := checkMinMaxValue(node.SplitOpt.Max, "max")
+	if err != nil {
+		return nil, err
+	}
+	p.Min = minValues
+	p.Max = maxValue
```
Comment (winkyao) [inline planner/core/planbuilder.go]: maxValues

File: planner/core/planbuilder.go (hunk 8 above)
Comment (crazycs520) [reply planner/core/planbuilder.go]: I will add some comment for this.

**Comment by winkyao** [comment]
@zimulala @bb7133 PTAL

File: executor/executor_test.go [hunk 10]
```diff
@@ -3801,6 +3801,32 @@ func (s *testSuite) TestSplitIndexRegion(c *C) {
 	c.Assert(err, NotNil)
 	terr := errors.Cause(err).(*terror.Error)
 	c.Assert(terr.Code(), Equals, terror.ErrCode(mysql.WarnDataTruncated))
+
+	// Check min value is more than max value.
+	tk.MustExec(`split table t index idx1 between (0) and (1000000000) regions 10`)
```
Comment (bb7133) [inline executor/executor_test.go]:
```
tidb> create table t(a int, b int, key b_idx(b));
Query OK, 0 rows affected (0.00 sec)

tidb> split table t index b_idx between (aa) and (ac) regions 10;
ERROR 1054 (42S22): Unknown column 'aa' in ''

tidb> split table t index b_idx between ('ac') and (ac) regions 10;
ERROR 1265 (01000): Data Truncated
```
Can we refine the error message?

File: executor/split.go [hunk 11]
```diff
@@ -75,10 +79,128 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+		binary.BigEndian.PutUint64(buf, startV)
+		value = append(value, buf...)
+		valuesList = append(valuesList, value)
+	}
+	return valuesList
+}
+
+// longestCommonPrefixLen gets the longest common prefix byte length.
+func longestCommonPrefixLen(s1, s2 []byte) int {
+	l := len(s1)
```
Comment (bb7133) [inline executor/split.go]: It's so weird that there's no builtin `min` function for integers in Go.

**Comment by bb7133** [comment]
Please update the parser dependency.

File: executor/executor_test.go (hunk 10 above)
Comment (crazycs520) [reply executor/executor_test.go]: Great catch! Thanks.

I think the error of `Unknown column` is reasonable.
```sql
root:test> create table t (a int, b int,index idx(a));
Query OK, 0 rows affected
Time: 0.065s
root:test> insert into t values (aa,bb);
(1054, u"Unknown column 'aa' in ''")
root:test>
```
I will try to refine the error of `Data Truncated`.
@bb7133

File: executor/split.go (hunk 11 above)
Comment (crazycs520) [reply executor/split.go]: done. I have used mathutil pkg.

**Comment by bb7133** [comment]
LGTM

File: planner/core/planbuilder.go [hunk 12]
```diff
@@ -1619,44 +1620,108 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+	lowerValues, err := checkLowerUpperValue(node.SplitOpt.Lower, "lower")
+	if err != nil {
+		return nil, err
 	}
-	return &SplitIndexRegion{
-		Table:      tableInPlan,
-		IndexInfo:  indexInfo,
-		ValueLists: indexValues,
-	}, nil
+	UpperValues, err := checkLowerUpperValue(node.SplitOpt.Upper, "upper")
```
Comment (zimulala) [inline planner/core/planbuilder.go]: s/UpperValues/upperValues? Like `lowerValues`.

File: executor/split.go [hunk 13]
```diff
@@ -75,10 +80,125 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+	for ; i < l; i++ {
+		if s1[i] != s2[i] {
+			break
+		}
+	}
+	return i
+}
+
+// getStepValue gets the step of between the lower and upper value. step = (upper-lower)/num.
+// convert byte slice to uint64 first.
```
Comment (zimulala) [inline executor/split.go]: s/convert/Convert

File: executor/split.go [hunk 14]
```diff
@@ -75,10 +80,125 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+		}
+		return nil, errors.Errorf("Split index region `%v` lower value %v should less than the upper value %v", e.indexInfo.Name, lowerStr, upperStr)
+	}
+	return getValuesList(lowerIdxKey, upperIdxKey, e.num, idxKeys), nil
+}
+
+// getValuesList is used to get `num` values between lower and upper value.
+// To Simplify the explain, suppose lower and upper value type is int64, and lower=0, upper=100, num=10,
+// then calculate the step=(upper-lower)/num=10, then the function should return 0+10, 10+10, 20+10... all together 9 (num-1) values.
+// then the function will return [10,20,30,40,50,60,70,80,90].
```
Comment (zimulala) [inline executor/split.go]: s/then/Then

File: executor/split_test.go [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+	// So 10 regions range is like below:
+	// region1: -inf ~ 10
+	// region2: 10 ~ 20
+	// region3: 20 ~ 30
+	// region4: 30 ~ 40
+	// region5: 40 ~ 50
+	// region6: 50 ~ 60
+	// region7: 60 ~ 70
+	// region8: 70 ~ 80
+	// region9: 80 ~ 80
```
Comment (zimulala) [inline executor/split_test.go]: s/80 ~ 80/80 ~ 90

File: executor/split.go [hunk 16]
```diff
@@ -75,10 +80,125 @@ func (e *SplitIndexRegionExec) Next(ctx context.Context, _ *chunk.RecordBatch) e
...
+			idxKeys = append(idxKeys, idxKey)
+		}
+		return idxKeys, nil
+	}
+	// Split index regions by lower, upper value and calculate the step by (upper - lower)/num.
+	lowerIdxKey, _, err := index.GenIndexKey(e.ctx.GetSessionVars().StmtCtx, e.lower, math.MinInt64, nil)
+	if err != nil {
+		return nil, err
+	}
+	upperIdxKey, _, err := index.GenIndexKey(e.ctx.GetSessionVars().StmtCtx, e.upper, math.MinInt64, nil)
```
Comment (zimulala) [inline executor/split.go]: Do we need to update `math.MinInt64` to `math.MaxInt64`?

File: executor/split_test.go [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+		Table:   model.NewCIStr("t1"),
+		Columns: idxCols,
+		State:   model.StatePublic,
+	}
+
+	// Test for int index.
+	// range is 0 ~ 100, and split into 10 region.
+	// So 10 regions range is like below:
+	// region1: -inf ~ 10
+	// region2: 10 ~ 20
```
Comment (zimulala) [inline executor/split_test.go]: Could you add symbols about open and closed intervals?

File: executor/split.go (hunk 16 above)
Comment (crazycs520) [reply executor/split.go]: Great, `math.MaxInt64` is better here.

File: executor/split_test.go (hunk 17 above)
Comment (crazycs520) [reply executor/split_test.go]: Good catch. 
done.

File: executor/split.go (hunk 16 above)
Comment (crazycs520) [reply executor/split.go]: Already add a comment here.

File: planner/core/planbuilder.go [hunk 18]
```diff
@@ -1619,44 +1620,108 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
+	schema := expression.TableInfo2SchemaWithDBName(b.ctx, node.Table.Schema, tblInfo)
+	mockTablePlan.SetSchema(schema)
 
-	indexValues := make([][]types.Datum, 0, len(node.ValueLists))
-	for i, valuesItem := range node.ValueLists {
-		if len(valuesItem) > len(indexInfo.Columns) {
-			return nil, ErrWrongValueCountOnRow.GenWithStackByArgs(i + 1)
-		}
-		valueList := make([]types.Datum, 0, len(valuesItem))
+	convertValue2ColumnType := func(valuesItem []ast.ExprNode) ([]types.Datum, error) {
```
Comment (zz-jason) [inline planner/core/planbuilder.go]: can you split this function outside the definition of `buildSplitRegion()`? It's a complicated function with 40 lines, which reduces the readability of `buildSplitRegion()`

**Pull request merged**